pkg/webui/account/views/validate/index.js

```diff
@@ -45,70 +46,74 @@ const m = defineMessages({
 const siteName = selectApplicationSiteName()
 const siteTitle = selectApplicationSiteTitle()
 
-export default class Validate extends React.PureComponent {
-  static propTypes = {
-    location: PropTypes.location.isRequired,
-  }
+const Validate = ({ location }) => {
+  const [error, setError] = useState(undefined)
+  const [success, setSuccess] = useState(undefined)
+  const [fetching, setFetching] = useState(true)
 
-  state = {
-    error: undefined,
-    success: undefined,
-    fetching: true,
-  }
+  const handleError = useCallback(error => {
+    setError(error)
+    setFetching(false)
+    setSuccess(undefined)
+  }, [])
 
-  handleError(error) {
-    this.setState({ error, fetching: false, success: undefined })
-  }
+  const handleSuccess = useCallback(() => {
+    setError(undefined)
+    setFetching(false)
+    setSuccess(m.validateSuccess)
+  }, [])
 
-  handleSuccess() {
-    this.setState({ success: m.validateSuccess, fetching: false, error: undefined })
-  }
-
-  async componentDidMount() {
-    const validationData = queryString.parse(this.props.location.search)
-    try {
-      await api.users.validate({
-        token: validationData.token,
-        id: validationData.reference,
-      })
-      this.handleSuccess()
-    } catch (error) {
-      if (isNotFoundError(error)) {
-        this.handleError(createFrontendError(m.tokenNotFoundTitle, m.tokenNotFoundMessage))
-      } else {
-        this.handleError(error)
+  useEffect(() => {
+    const makeRequest = async () => {
+      const validationData = queryString.parse(location.search)
+      try {
+        await tts.ContactInfo.validate({
+          token: validationData.token,
+          id: validationData.reference,
+        })
+        handleSuccess()
+      } catch (error) {
+        if (isNotFoundError(error)) {
+          handleError(createFrontendError(m.tokenNotFoundTitle, m.tokenNotFoundMessage))
+        } else {
+          handleError(error)
+        }
       }
     }
-  }
+    makeRequest()
```

Comment:
Any reason for extracting the whole logic to a local function?

Reply:
Because we need to run async logic there and it's not allowed to use an async function directly as argument for `useEffect()`, since the effect function must either return a cleanup function or nothing (but would then return a promise).